---
title: "MCMC report"
output: github_document
params: 
  mcmc_output: 1
  mcmc_output_fixed_epsilon: 1
  desman_input_file: 1
  fasta_file: 1
  variants: 1
---





# Setup

## Project ID

` r Sys.getenv("TAR_PROJECT")`

## Input data
The counts per position, sample and nucleotides array was obtained from:
`r desman_input_file`
```{r,eval=TRUE,echo=FALSE}
knitr::asis_output(if(exists(fasta_file)){paste0("
The fasta file used for the dictionnary was: [",fasta_file,"]")
}else{
  paste0("An ad hoc dictionnary of size ",length(variants)," was used")
})
```

## MCMC Tuning parameters


# Results

```{r}
plot_bar_epsilon(mcmc_output)
```

# Reproduction

To reproduce the code:
If not done, clone the repo: 
[https://github.com/konkam/DesmanResultsAnalysis]

then run:

```{r,eval=TRUE,echo=FALSE}
knitr::asis_output(paste0("
``` r
library(targets)
Sys.setenv(TAR_PROJECT = '",Sys.getenv("TAR_PROJECT"),"')
tar_make()
```"))
```
